### DISCUSSION TITLE
Command Substitution & Error Codes

### DISCUSSION THREAD

**Discussion by prodigy-byron-jones** [go-task/task#687] [Q&A] [answered]
Hi all,

First time poster here and I have to say this is an incredibly versatile and powerful tool. Thank you for all the hard work!

That being said I am incredibly confused by how command substitution (and I suppose by extension sub-shells) are handled by task .

If I run the following bash, I see the expected error code output of 255
```
test=$(exit 255)
echo $?
```

The same equivalent in task always returns a 0 error code:
```
  test:
    cmds:
      - |
        test=$(exit 255)
        echo $?
```  

Do I have some gap in my knowledge or is this behaviour entirely unexpected?
Is there a way for me to correctly do this without having to reference an external bash file?

Thank you :)

**Comment by ghostsquad** [comment]
What happens if you add a semicolon after line 1?

What version of task are you running?

Generally though, this may be an issue best directed at https://github.com/mvdan/sh for help. This is what is used under the hood of task.

**Reply by prodigy-byron-jones** [reply]
Thank you for the response @ghostsquad.

A semicolon after line 1 still unfortunately returns the same result of 0.

I am making use of the following version:
```
Task version: v3.11.0 (h1:eCcCe1cKyta4YnGQYLo2FriPS9h7fFS8XBVW6QvWEV4=)
```

Thanks for the tip about the underlying shell parser. I am quite surprised no other task users have experienced/logged this issue aha.

**Reply by ghostsquad** [reply]
I can reproduce:

```
at 10:21:56 ❯ t test
task: [test] test=$(exit 255)
echo $?

0
```

There's probably a use case to allow a user to specify a different shell, and use the system's shell. This would possibly mean an inconsistent behavior depending on host, with `mvdan/sh` being the default.

Honestly, I wouldn't recommend doing very much "bash-specific" anything with task, besides calling single commands or scripts.

**Reply by prodigy-byron-jones** [reply]
Heya @ghostsquad, I would say this is the exact use case of Task :)

Good helper scripts handle redirecting output and handling errors effectively even if they are small. Shifting out into an external shell file for small/medium scripts would be a little bit redundant and seemingly antithetical to the spirit of Task. For bigger items I couldn't agree more and I have done so in the past.

`mvdan/sh` seems to fit the bill perfectly and I don't think multiple shell parsers should be allowed (for complexity concerns). I for one would just love the bug to be sorted but understand this is not a problem with Task but rather one layer down.

I appreciate the feedback. Thank you!

**Comment by andreynering** [comment]
Hi @prodigy-byron-jones,

I have the impression that this may be a by design decision of [the shell library](https://github.com/mvdan/sh) to not support some advanced things like this one.

I opened an issue to check: https://github.com/mvdan/sh/issues/828.

**Reply by prodigy-byron-jones** [reply]
Hiya @andreynering,

Thank you so much for the response and logging the issue (and for Task :D). My understanding would be that they want to follow Bash as closely as possible. This is seeming to me more and more like it is an actual bug in `mvdan/sh` :)

**Comment by prodigy-byron-jones** [comment, marked as answer]
For those who check this discussion. There was indeed a bug with `mvdan/sh` that consequently has been patched. In theory if the lib is updated with the next build of Task it should be resolved for us.
